util/ranger/detacher.go

```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	record := make([]types.Datum, 1)
+
+	expr := cols[0].VirtualExpr.Clone()
+	andType := types.NewFieldType(mysql.TypeTiny)
+
+	sf := accessesCond[1].(*expression.ScalarFunction)
+	c, _ := sf.GetArgs()[0].(*expression.Column)
+	var AndOrExpr expression.Expression
+	for i, arg := range sf.GetArgs()[1:] {
+		// get every const value and calculate tidb_shar(val)
+		con, _ := arg.(*expression.Constant)
+		conVal, err := con.Eval(chunk.Row{})
+		if err != nil {
+			return accessesCond, err
+		}
+
+		record[0] = conVal
+		mutRow := chunk.MutRowFromDatums(record)
+		exprVal, err := expr.Eval(mutRow.ToRow())
+		if err != nil {
+			return accessesCond, err
+		}
+
+		// tmpArg1 is like `tidb_shard(a) = 8`, tmpArg1 is like `a = 100`
+		exprCon := &expression.Constant{Value: exprVal, RetType: cols[0].RetType}
+		tmpArg1, err1 := expression.NewFunction(sctx, ast.EQ, cols[0].RetType, cols[0], exprCon)
+		if err1 != nil {
+			return accessesCond, err1
+		}
+		tmpArg2, err2 := expression.NewFunction(sctx, ast.EQ, c.RetType, c.Clone(), arg)
+		if err2 != nil {
+			return accessesCond, err2
+		}
+
+		// make a LogicAnd, e.g. `tidb_shard(a) = 8 AND a = 100`
+		andExpr, err3 := expression.NewFunction(sctx, ast.LogicAnd, andType, tmpArg1, tmpArg2)
+		if err3 != nil {
+			return accessesCond, err3
+		}
+
+		if i == 0 {
+			AndOrExpr = andExpr
+		} else {
+			// if the LogicAnd more than one, make a LogicOr,
+			// e.g. `(tidb_shard(a) = 8 AND a = 100) OR (tidb_shard(a) = 161 AND a = 200)`
+			AndOrExpr, errRes = expression.NewFunction(sctx, ast.LogicOr, andType, AndOrExpr, andExpr)
+			if errRes != nil {
+				return accessesCond, errRes
+			}
+		}
+	}
+
+	newAccessCond = append(newAccessCond, AndOrExpr)
+
+	return newAccessCond, nil
+}
+
+// AddGcColumn4EqCond add the `tidb_shard(x) = xxx` prefix for equal condition
+// For param explanation, please refer to the function `AddGcColumnCond`.
+// @retval -  []expression.Expression   the new condtions after adding `tidb_shard() = xxx` prefix
+//            []*valueInfo              the values of every columns in the returned new condtions
+//            error                     if error gernerated, return error
+func AddGcColumn4EqCond(sctx sessionctx.Context,
+	cols []*expression.Column,
+	accessesCond []expression.Expression,
+	columnValues []*valueInfo) ([]expression.Expression, error) {
+
+	expr := cols[0].VirtualExpr.Clone()
+	record := make([]types.Datum, len(columnValues)-1)
+
+	for i := 1; i < len(columnValues); i++ {
+		cv := columnValues[i]
+		if cv == nil {
+			break
+		}
+		record[i-1] = *cv.value
+	}
+
+	mutRow := chunk.MutRowFromDatums(record)
+	evaluated, err := expr.Eval(mutRow.ToRow())
+	if err != nil {
+		return accessesCond, err
+	}
+	vi := &valueInfo{false, &evaluated}
+	con := &expression.Constant{Value: evaluated, RetType: cols[0].RetType}
+	// make a tidb_shard() function, e.g. `tidb_shard(a) = 8`
+	cond, err := expression.NewFunction(sctx, ast.EQ, cols[0].RetType, cols[0], con)
```

Comment:
@you06 @cfzjywxk PTAL again!
I replace NewFunctionInternal to NewFunction that can return an error if something failed.